### DISCUSSION TITLE
Why some operations are few folds slower compared to CUDA and DPC++

### DISCUSSION THREAD

**Discussion by KestutisMa** [AdaptiveCpp/AdaptiveCpp#1491] [Q&A]
First of all thank you for your efforts developing such great tool and keeping it open source.
I tried to read any comparisons of AdaptiveCPP, DPC++ and CUDA in scientific literature and first came upon this, could you comment a little bit about their results? [Lessons Learned Migrating CUDA to SYCL: A HEP Case Study with ROOT RDataFrame](https://arxiv.org/html/2401.13310v1)
Thanks again, keep it going.

**Comment by illuhad** [comment]
Hi, my knowledge about this particular code is very limited, so I cannot give you in depth explanations why every single result is the way it is. For details, I would recommend reaching out to Jolly, the first author, directly.

I was there when these results were presented at IWOCL 24. Jolly and I (and others) were trying to make sense of what we did see. For example, we don't really understand why AdaptiveCpp has so many more CUDA runtime API calls, which does not really make sense. I think this indicates that there likely something special about the code or the methodology, so that some care has to be taken interpreting these results - perhaps we are not looking at the same thing between implementations. 
 But as I did not look at the code, I cannot tell you what this might be.

Unlike DPC++, AdaptiveCpp by default uses background worker threads for deferred submission, and one thing we suspected is that e.g. overlapped API calls in the multi-threaded AdaptiveCpp runtime is probably not taken into account in the visualization of fig. 4, creating a misleading impression that AdaptiveCpp might slow down the code with unneeded API calls. In other words, fig 4 and 5 might show stacked times for things that perhaps actually run in parallel.

One thing I can say is that this work used our old SMCP compiler, not our newer generic SSCP JIT compiler, which is known to produce substantially faster binaries and which we recommend to users. Also, this application heavily uses SYCL 2020 reductions, and those might have a substantial impact on performance. Reductions were recently completely reimplemented with multiple improvements, so the current reduction implementation will have very different performance characteristics compared to what is shown in the paper.

**Reply by illuhad** [reply]
PS: Here you can find some newer results that show how our newer generic JIT compiler (the default compiler that was not yet used in this paper) compares against DPC++: https://github.com/AdaptiveCpp/AdaptiveCpp/releases/tag/v24.02.0

Also keep in mind that DPC++ by default compiles with `-ffast-math` and approximate square root, whereas AdaptiveCpp tries to be accurate by default unless switches like `-ffast-math` are provided. So, to obtain comparable results, you would need to be aware of these differences and account for them by using potentially different sets of compilation flags. I don't know to what extent this was done in this paper. My feeling is however that many comparisons that you can find in literature, this is often not taken into account.
